### PR TITLE
Added support for marking functions as Leaf.

### LINKED ISSUE #238: Error outputs are not understood.
Running the ffigen on LibGit2's Source, the output produces many Warnings and a Severe. The issue is that there is no documentation (at least that I can find) as to what what they mean and what to do about them. Are these errors okay?

Do skipped functions need to be manually created? Do Generated declarations need to be modified? Is there documentation on "unsupported return type or parameter type" and why? Documentation on what types are unsupported?

If skipped functions do need attention, it would be cool if the warnings showed the file they were in.

It produces the following:
```
Input Headers: [./libgit2/include/git2.h]
[SEVERE] : Header ./libgit2/include/git2.h: Total errors/warnings: 1.
[SEVERE] :     ./libgit2/include/git2/sys/credential.h:10:10: fatal error: 'git2/common.h' file not found [Lexical or Preprocessor Issue]
[WARNING]: Skipped Function 'strtold', function has unsupported return type or parameter type.
[WARNING]: Skipped Function '__bswap_16', inline functions are not supported.
[WARNING]: Skipped Function '__bswap_32', inline functions are not supported.
[WARNING]: Skipped Function '__bswap_64', inline functions are not supported.
[WARNING]: Skipped Function '__uint16_identity', inline functions are not supported.
[WARNING]: Skipped Function '__uint32_identity', inline functions are not supported.
[WARNING]: Skipped Function '__uint64_identity', inline functions are not supported.
[WARNING]: Skipped Function 'qecvt', function has unsupported return type or parameter type.
[WARNING]: Skipped Function 'qfcvt', function has unsupported return type or parameter type.
[WARNING]: Skipped Function 'qgcvt', function has unsupported return type or parameter type.
[WARNING]: Skipped Function 'qecvt_r', function has unsupported return type or parameter type.
[WARNING]: Skipped Function 'qfcvt_r', function has unsupported return type or parameter type.
[WARNING]: Generated declaration '__fsid_t' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__locale_data' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__locale_struct' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__tzname' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__daylight' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__timezone' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__ctype_get_mb_cur_max' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__sigset_t' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__pthread_list_t' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__pthread_slist_t' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__pthread_mutex_s' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__pthread_rwlock_arch_t' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__pthread_cond_s' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_Exit' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_LIBSSH2_USERAUTH_KBDINT_PROMPT' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_LIBSSH2_USERAUTH_KBDINT_RESPONSE' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_LIBSSH2_SESSION' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_TIME_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_FEATURES_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_DEFAULT_SOURCE' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__USE_ISOC11' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__USE_ISOC99' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__USE_ISOC95' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_POSIX_SOURCE' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_POSIX_C_SOURCE' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__USE_POSIX' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__USE_POSIX2' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__USE_POSIX199309' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__USE_POSIX199506' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__USE_XOPEN2K' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__USE_XOPEN2K8' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_ATFILE_SOURCE' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__USE_MISC' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__USE_ATFILE' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__USE_FORTIFY_LEVEL' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__GLIBC_USE_DEPRECATED_GETS' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_STDC_PREDEF_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__STDC_IEC_559__' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__STDC_IEC_559_COMPLEX__' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__STDC_ISO_10646__' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__GNU_LIBRARY__' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__GLIBC__' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__GLIBC_MINOR__' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_SYS_CDEFS_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__glibc_c99_flexarr_available' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__WORDSIZE' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__WORDSIZE_TIME64_COMPAT32' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__SYSCALL_WORDSIZE' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__HAVE_GENERIC_SELECTION' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_BITS_TIME_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_BITS_TYPES_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_BITS_TYPESIZES_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__OFF_T_MATCHES_OFF64_T' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__INO_T_MATCHES_INO64_T' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__RLIM_T_MATCHES_RLIM64_T' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__FD_SETSIZE' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__clock_t_defined' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__time_t_defined' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__struct_tm_defined' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_STRUCT_TIMESPEC' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__LITTLE_ENDIAN' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__BIG_ENDIAN' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__PDP_ENDIAN' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__BYTE_ORDER' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__FLOAT_WORD_ORDER' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__clockid_t_defined' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__timer_t_defined' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__itimerspec_defined' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_BITS_TYPES_LOCALE_T_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_BITS_TYPES___LOCALE_T_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__GLIBC_USE_LIB_EXT2' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__GLIBC_USE_IEC_60559_BFP_EXT' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__GLIBC_USE_IEC_60559_FUNCS_EXT' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__GLIBC_USE_IEC_60559_TYPES_EXT' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_STDLIB_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__WNOTHREAD' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__WALL' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__WCLONE' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__ENUM_IDTYPE_T' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__W_CONTINUED' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__WCOREFLAG' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__HAVE_FLOAT128' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__HAVE_DISTINCT_FLOAT128' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__HAVE_FLOAT64X' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__HAVE_FLOAT64X_LONG_DOUBLE' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__HAVE_FLOAT16' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__HAVE_FLOAT32' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__HAVE_FLOAT64' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__HAVE_FLOAT32X' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__HAVE_FLOAT128X' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__HAVE_DISTINCT_FLOAT16' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__HAVE_DISTINCT_FLOAT32' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__HAVE_DISTINCT_FLOAT64' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__HAVE_DISTINCT_FLOAT32X' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__HAVE_DISTINCT_FLOAT64X' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__HAVE_DISTINCT_FLOAT128X' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__HAVE_FLOATN_NOT_TYPEDEF' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__ldiv_t_defined' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__lldiv_t_defined' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_SYS_TYPES_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_BITS_STDINT_INTN_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__BIT_TYPES_DEFINED__' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_ENDIAN_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_BITS_BYTESWAP_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_BITS_UINTN_IDENTITY_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_SYS_SELECT_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__FD_ZERO_STOS' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__sigset_t_defined' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_SIGSET_NWORDS' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__timeval_defined' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__NFDBITS' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_BITS_PTHREADTYPES_COMMON_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_THREAD_SHARED_TYPES_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_BITS_PTHREADTYPES_ARCH_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__SIZEOF_PTHREAD_MUTEX_T' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__SIZEOF_PTHREAD_ATTR_T' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__SIZEOF_PTHREAD_RWLOCK_T' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__SIZEOF_PTHREAD_BARRIER_T' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__SIZEOF_PTHREAD_MUTEXATTR_T' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__SIZEOF_PTHREAD_COND_T' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__SIZEOF_PTHREAD_CONDATTR_T' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__SIZEOF_PTHREAD_RWLOCKATTR_T' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__SIZEOF_PTHREAD_BARRIERATTR_T' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__PTHREAD_MUTEX_HAVE_PREV' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__PTHREAD_RWLOCK_ELISION_EXTRA' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__have_pthread_attr_t' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_ALLOCA_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_INTTYPES_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_STDINT_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_BITS_WCHAR_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__WCHAR_MAX' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__WCHAR_MIN' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '_BITS_STDINT_UINTN_H' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '____gwchar_t_defined' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__PRI64_PREFIX' start's with '_' and therefore will be private.
[WARNING]: Generated declaration '__PRIPTR_PREFIX' start's with '_' and therefore will be private.
```

### LINKED ISSUE #248: test/debug_generated/README.md
```
Package validation found the following potential issue:
* 1 checked in file is ignored by a `.gitignore`.
  Previous versions of Pub would include those in the published package.
  
  Consider adjusting your `.gitignore` files to not ignore those files, and if you do not wish to
  publish these files use `.pubignore`. See also dart.dev/go/pubignore
  
  Files that are checked in while gitignored:
  
  test/debug_generated/README.md
```

Not urgent, but we should probably clean it up.

Possibly just remove this file (which removes the folder in git) and then make sure that the tests will create the folder if they try to write here.

### LINKED ISSUE #251: Add an option to use leaf functions for C functions that don't accept callbacks
I was reading [https://gist.github.com/mraleph/2582b57737711da40262fad71215d62e](https://gist.github.com/mraleph/2582b57737711da40262fad71215d62e) and it mentioned that calling a leaf function(NativeFunction.asFunction(leaf: true)) is faster than calling a non-leaf function and might at some point allow the C function to receive pointers into Dart memory such as typed data without having to allocate temporary C memory.
There should be an option for making leaf functions if they don't have any callback parameters. It should probably not be the default because the C function might still have a Dart function pointer saved somewhere.

### PR DESCRIPTION
**Pull request by mannprerak2**
Closes #251, Closes #248, Closes #238

- Added support for passing `isLeaf:true` to functions.
- Added/updated tests.
- Update version, changelog, readme.
- Removed the debug_generated folder, tests will now automatically create that folder upon running.
- Add an FAQ to readme regarding generated logs.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -596,3 +618,16 @@ The following typedefs are not generated -
 - They are not referred to anywhere in the included declarations.
 - They refer to a struct/union having the same name as itself.
 - They refer to a boolean, enum, inline array, Handle or any unsupported type.
+
+### What are these logs generated by ffigen and how to fix them?
+
+Ffigen can sometimes generate a lot of logs, especially when it's parsing a lot of code.
+- `SEVERE` logs are something you *definitely need to address*. They can be caused due
+to syntax errors, or more generally missing header files (which need to be specified
```
Comment (kevmoo) [inline README.md:626]: please indent the markdown here so it's easier to read when not rendered

Comment (mannprerak2) [reply README.md]: Done, Is this fine?

File: README.md [hunk 2]
```diff
@@ -593,6 +615,19 @@ that refers to them.
 ### Why are some typedefs not generated?
 
 The following typedefs are not generated -
-- They are not referred to anywhere in the included declarations.
-- They refer to a struct/union having the same name as itself.
-- They refer to a boolean, enum, inline array, Handle or any unsupported type.
+  - They are not referred to anywhere in the included declarations.
+  - They refer to a struct/union having the same name as itself.
+  - They refer to a boolean, enum, inline array, Handle or any unsupported type.
```
Comment (dcharkes) [inline README.md:620]: Why indent these, but not indent the ones under the next header?

File: README.md [hunk 3]
```diff
@@ -593,6 +615,19 @@ that refers to them.
...
+Ffigen can sometimes generate a lot of logs, especially when it's parsing a lot of code.
+- `SEVERE` logs are something you *definitely need to address*. They can be caused due
+  to syntax errors, or more generally missing header files (which need to be specified
+  using `compiler-opts` in config)
+- `WARNING` logs are something *you can ignore*, but should probably look into. These are mostly
+  indications of declarations ffigen couldn't generate due to limitations of dart:ffi,
+  private declarations (which can be resolved by renaming them via ffigen config) or other minor issues
+  in the config file itself.
+- Everything else can be safely ignored. It's purpose is to simply
+  let you know what ffigen is doing.
```
Comment (dcharkes) [inline README.md:633]: Maybe add a `-` with how to change the verbosity.

**Review by dcharkes — APPROVED** [review]
lgtm

cc @cskau-g We're now going to generate `isLeaf` FFI calls from ffigen! 🚀

**Comment by cskau-g** [comment]
> * Added support for passing `isLeaf:true` to functions.

Fantastic! Thank you.

File: README.md (hunk 2 above)
Comment (mannprerak2) [reply README.md:620]: Fixed.

File: README.md (hunk 3 above)
Comment (mannprerak2) [reply README.md]: Done

**Comment by mannprerak2** [comment]
@dcharkes I've made the required changes.

**Pull request merged**